src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+                                     IDesignTimeInputsDataSource inputsDataSource,
+                                     IDesignTimeInputsFileWatcher fileWatcher,
+                                     ITempPECompiler compiler,
+                                     IFileSystem fileSystem)
+            : base(threadingService.JoinableTaskContext)
+        {
+            _project = project;
+            _projectSubscriptionService = projectSubscriptionService;
+            _activeWorkspaceProjectContextHost = activeWorkspaceProjectContextHost;
+            _inputsDataSource = inputsDataSource;
+            _fileWatcher = fileWatcher;
+            _compiler = compiler;
+            _fileSystem = fileSystem;
+
+            _scheduler = new TaskDelayScheduler(s_compilationDelayTime, threadingService, CancellationToken.None);
+        }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Create an action block process the design time inputs and configuration general changes
+            _inputsActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<Tuple<DesignTimeInputs, IProjectSubscriptionUpdate>>>(ProcessDataflowChanges);
+
+            IDisposable projectLink = ProjectDataSources.SyncLinkTo(
+                   _inputsDataSource.SourceBlock.SyncLinkOptions(
+                       linkOptions: new StandardRuleDataflowLinkOptions
+                       {
+                           PropagateCompletion = true,
+                       }),
+                   _projectSubscriptionService.ProjectRuleSource.SourceBlock.SyncLinkOptions(
+                      linkOptions: new StandardRuleDataflowLinkOptions
+                      {
+                          RuleNames = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName),
+                          PropagateCompletion = true,
+                      }),
+                   _inputsActionBlock,
+                   DataflowOption.PropagateCompletion,
+                   cancellationToken: _project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+
+            // Create an action block to process file change notifications
+            _fileWatcherActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<string[]>>(ProcessFileChangeNotification);
+            IDisposable watcherLink = _fileWatcher.SourceBlock.LinkTo(_fileWatcherActionBlock, DataflowOption.PropagateCompletion);
+
+            _disposables.AddDisposable(projectLink);
+            _disposables.AddDisposable(watcherLink);
+
+            return Task.CompletedTask;
+        }
+
+        internal void ProcessFileChangeNotification(IProjectVersionedValue<string[]> arg)
+        {
+            // Ignore any updates until we've received the first set of design time inputs (which shouldn't happen anyway)
+            // That first update will queue all of the files so we're not losing anything
```

Comment:
Why should this never happen?

Reply:
It _shouldn't_, but as you say in a below comment, the calls could overlap so a bit of extra protection didn't seem like a terrible idea.